upup/pkg/fi/cloudup/gce/tpm/gcetpmsigner/tpmauthenticator.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcetpmsigner
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/go-tpm-tools/client"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/pkg/apis/nodeup"
+	"k8s.io/kops/pkg/bootstrap"
+)
+
+type tpmAuthenticator struct {
+	projectID string
+	zone      string
+	instance  string
+}
+
+var _ bootstrap.Authenticator = &tpmAuthenticator{}
+
+func NewTPMAuthenticator() (bootstrap.Authenticator, error) {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil, fmt.Errorf("error getting projectID from metadata: %w", err)
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil, fmt.Errorf("error getting zone from metadata: %w", err)
+	}
+	instance, err := metadata.InstanceName()
+	if err != nil {
+		return nil, fmt.Errorf("error getting instance from metadata: %w", err)
+	}
+
+	return &tpmAuthenticator{
+		projectID: projectID,
+		zone:      zone,
+		instance:  instance,
+	}, nil
+}
+
+func (a *tpmAuthenticator) SignBootstrapRequest(req *nodeup.BootstrapRequest) (*bootstrap.SignedRequestData, error) {
+	tpmStart := time.Now()
+
+	tpmDevice, err := openTPM()
+	if err != nil {
+		return nil, fmt.Errorf("failed to open TPM: %w", err)
+	}
+	defer tpmDevice.Close()
+
+	key, err := client.GceAttestationKeyRSA(tpmDevice)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get GCE RSA attestation key from TPM: %w", err)
+	}
+	defer key.Close()
+
+	klog.V(2).Infof("attestation key is %v", debugToPEM(key.PublicKey()))
+
+	klog.Infof("TPM initialization took %v", time.Since(tpmStart))
+
+	if req.Claim == nil {
+		req.Claim = &nodeup.Claim{}
+	}
+
+	{
+		node := &nodeup.NodeClaim{
+			GCPProjectID: a.projectID,
+			Zone:         a.zone,
+			Instance:     a.instance,
+		}
+
+		payload, err := json.Marshal(node)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal node claim: %w", err)
+		}
+		signature, err := tpmSign(key, payload)
+		if err != nil {
+			return nil, fmt.Errorf("failed to sign node claim: %w", err)
```

Comment:
Why sign the node claim? What would an attacker gain by altering it?

Reply:
Changed the way the signature matches to match the way it works on the AWS implementation.

Previously, we were demonstrating that we had control of the signing key, but it's a lot more prone to subtle mistakes ... as you spotted :-)